Add SPACEXLAUNCH as a supported price identifier (#295)
UMAprotocol/UMIPs: UMIPs/spacex_launch_price_identifier.md

## Headers

| UMIP-XX                |                                                               |
| ------------------- | ------------------------------------------------------------- |

> | UMIP Title          | Add SPACEXLAUNCH as a supported price identifier |

Reviewer: Very very interesting idea

| Authors             | Sean Brown (@smb2796), Andrey Belyakov (andrey@opium.team), Ali Nuraldin (ali@opium.team) |
| Status              | Draft                                                         |
| Created             | May 19, 2021                                              |
| Discourse Link      | [Link](https://discourse.umaproject.org/t/add-spacexlaunch-as-a-supported-price-identifier/1105)            |

# Summary 

> The DVM should support price requests for SPACEXLAUNCH. The purpose of the SPACEXLAUNCH price identifier is to be a generalized way to evaluate the success rate of the rocket launches made by SpaceX till date considering provided ancillary data

Reviewer 2: Until a specified date? Is the number a cumulative score increasing over time with the number of SpaceX launches? Would the price identifier consider all of the launches prior to the creation of the identifier? If so, it might be useful to show the current score along with a table of SpaceX's launch history matched against their score and weight, as an example of how the identifier works.

> The DVM should support price requests for SPACEXLAUNCH. The purpose of the SPACEXLAUNCH price identifier is to be a generalized way to evaluate the success rate of the rocket launches made by SpaceX till date considering provided ancillary data

Reviewer 3: ```suggestion
The DVM should support price requests for SPACEXLAUNCH. The purpose of the SPACEXLAUNCH price identifier is to be a generalized way to evaluate the success rate of the rocket launches made by SpaceX till date considering provided ancillary data.
```

Ancillary data in the request will specify the parameters that should be considered during the oracle result determination process.

> # Motivation

Reviewer 4: I think it is worth noting that this price identifier is not intended to be used with a registered UMA financial contract type. Is it possible to link to the Opium contract that will use this price id?

> Provision of the way to evaluate the success rate of SpaceX launches would open a wide range of possible synthetics based on it (e.g. insurances)

Reviewer 2: `Adding a way to evaluate the success rate of SpaceX launches would open a wide range of potential synthetics (e.g. insurance).`

# Data Specification

- Price identifier name: **SPACEXLAUNCH** 
- Markets & Pairs: **N/A**

> - Example data providers: https://www.rocketlaunch.live/?filter=spacex , https://en.wikipedia.org/wiki/List_of_Falcon_9_and_Falcon_Heavy_launches

Reviewer: How do we verify the authority of the source data? These websites could miss data, or report false data.

Reviewer 2: It would be great to have a few more data sources listed here, if possible, and industry sources to fact check Wikipedia and the like in the event of a dispute.

Reviewer 5: I think that multiple sources are definitely needed for repudiation.
Here is one that I found: https://docs.spacexdata.com/

- Cost to use: **None**
- Real-time data update frequency: **Daily**

> - Historical data update frequency: **Daily**

Reviewer 4: Bump: @Alirun can you update this to `NA`?

# Price Feed Implementation

> Price Feed Implementation is unnecessary for this price identifier, as it will not have liquidation bots that need to programmatically get this price.

Reviewer 2: Will this be calculated by hand on the expiry of a contract?

# Technical Specifications

- Price identifier name: **SPACEXLAUNCH**
- Base Currency: **N/A**
- Quote Currency: **N/A**

> - Rounding: *Round to 2 decimal places (third decimal place digit >= 5 rounds up and < 5 rounds down)*

Reviewer: What would this rounding do? Is this the percentage of launch success rate?

Reviewer 4: Rounding is still part of the template, so it seems fine to me.

- Estimated current value of price identifier: **N/A**

---

## Description

> Ancillary data would consist of the list of launches, where each launch is described by

Reviewer: How do we determine the list? Do we always use the beginning of time as the start filter date, and contract expiration as the end filter date?

Reviewer 4: Good questions! The list is passed in an ancillary data and Ali also updated to clarify that the end timestamp is the price request timestamp

- Launch id - Name of the SpaceX mission

> - Weight (`Wi`) - Weight of the launch in the overall success rate calculation

Reviewer 6: Whats the difference between Wi and Si?

> Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions

Reviewer: What about launches that didn't intend for the rocket to land? I feel like this ancillary data requires a review from the rocket domain expert. Where did you derive this success rate formula `Success Rate =  SUM(Si * Wi) / SUM(Wi)`?

Reviewer 4: I think it is a good point, but I see this as being more of a product question for the contract deployer around what launches are included in the ancillary data list. I don't think this has any impact on our ability to return a price as specified though.

> Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions

Reviewer 7: The formula and conditions for `Si` calculation seems more belonging to the Implementation section

Reviewer 4: Bump. @Alirun can you move the description/examples to the implementation section?

- If launch did not happen till date or the rocket was not successfully lifted off, then `Si=0`

> - If the rocket successfully lifted off, but never landed, or the landing was not successful then `Si=0.5`

Reviewer: Again, questioning why 0.5? The launch is difficult, and some may not intend to land.

Reviewer 6: If it lands on earth, or anywhere?  What if it docks at a space station, is that considered "landing"?

Reviewer 4: Agreed that the criteria for each bucket could use some additional clarification

> - If the rocket successfully lifted off, and landed, then `Si=1`

Reviewer 4: @Alirun can you also add more explicit language around this criteria in relation to the price request timestamp, like @Reinis-FRP said? Simply a "landing was not successful before the price request timestamp", and "liftoff was not successful before the price request timestamp"

> Liftoff is considered successful if the rocket reached expected altitude without any significant damage and performed the required part of the mission at this altitude.

Reviewer: We need to further define `expected altitude` as the rockets do see explosions even after reaching significant altitude. All these definitions should link a reference the authoritative source.

Reviewer 8: Agree with this comment. The rocket will reach its expected altitude and then be deemed a success but then blow up later, which for an insurance idea might not be useful. I am also not sure what significant damage means?

> Landing is considered successful if the rocket landed on Earth without any significant damage till date.

Reviewer: Why limited on `Earth`?

Success rate than can be calculated by
```

> Success Rate =  SUM(Si * Wi) / SUM(Wi)

Reviewer 2: Just to check my understanding, `Wi` would be passed in through ancillary data, and what the price proposer will need to provide is the result of the calculation after adding `Si` for each of the relevant launches?

Reviewer 4: This matches my understanding

```

i.e. sum of all products of launch statuses and corresponding weights divided by the sum of all weights

> ## Examples

Reviewer 7: I believe Examples belong more to the Implementation section

### Example 1

- Request timestamp: 1614556800 (March 1st 2021 00:00 UTC)
- List of launches:

> - - Launch 1: ID: "Starlink-18”, W: 1

Reviewer: You mean `Wi`?

Reviewer 2: I think `i` is the index (so `1` here and `2` for the next one).

Reviewer: Make sense

- - Launch 2: ID: "Starlink-19”, W: 1

> Launch 1 happened till date, successfully lifted off and landed: `S1=1`

Reviewer 2: `Launch 1 happened, successfully lifted off, and landed`

> Launch 2 happened till date, successfully lifted off, but did not landed: `S1=0.5`

Reviewer 2: `S2=0.5`, I think.

Reviewer 2: `Launch 2 happened, successfully lifted off, but did not land`

Calculating the success rate:
```
Success Rate = (1 * 1 + 1 * 0.5) / (1 + 1) = 0.75
```

### Example 2
- Request timestamp: 1612137600 (February 1st 2021 00:00 UTC)
- List of launches:

> - -  Launch 1: ID: "Transporter-1”, W: 1

Reviewer: `Wi`

> Launch 1 happened till date, successfully lifted off and landed: S1=1

Reviewer 2: `Launch 1 happened, successfully lifted off, and landed`

Calculating the success rate:
```
Success Rate = (1 * 1) / (1) = 1
```

##  Format

> Ancillary data in the request will be passed as a JSON  in the following format

Reviewer 2: Extra space here.

```
[{

>   id: string,

Reviewer: I can foresee future where we want to update different rocket type with different parameters. Not all launches are equal.

Reviewer 2: I think that's what the `weight` field is for, so you can mark certain launches as more or less important to your calculation.

Reviewer: My understanding was `weight` was used for the actual `weight of the rocket` lol. Fair, we should have a clear definition of the weight in this case

  weight: number
}]
```

# Rationale

> No rationale is needed. The motivation for this price identifier is explained in motivation.

Reviewer: I would argue rationale for the definition, and formula selection should be included here. As a first time reader, I'm not convinced this is the right implementation (although I agree this is a very very innovative idea). Let's cite some sources where we get these formula

> No rationale is needed. The motivation for this price identifier is explained in motivation.

Reviewer 4: Agreed. The results of this price are manipulatable without some criteria around the weighting associated with each launch.

> No rationale is needed. The motivation for this price identifier is explained in motivation.

Reviewer 4: bump: @Alirun can you add some clarification around what how "weights" are chosen? This isn't incredibly necessary, but would help voters determine when ancillary data requests were submitted incorrectly.

    buyerPayout = _derivative.margin.sub(sellerPayout);
  } else {
    buyerPayout = 0;
    sellerPayout = _derivative.margin;
  }

  // Add premium
  sellerPayout = sellerPayout.add(fixedPremium);
}
```

Price requester should consider additional time buffers for potential launch delays on it's own.

# Data Specification

- Price identifier name: **SPACEXLAUNCH** 
- Markets & Pairs: **N/A**
- Example data providers: https://www.rocketlaunch.live/?filter=spacex , https://en.wikipedia.org/wiki/List_of_Falcon_9_and_Falcon_Heavy_launches , https://docs.spacexdata.com/
- Cost to use: **None**
- Real-time data update frequency: **Daily**
- Historical data update frequency: **Daily**

# Price Feed Implementation

Price Feed Implementation is unnecessary for this price identifier, as it will not have liquidation bots that need to programmatically get this price.

# Technical Specifications

- Price identifier name: **SPACEXLAUNCH**
- Base Currency: **N/A**
- Quote Currency: **N/A**
- Estimated current value of price identifier: **N/A**

---

## Description

Ancillary data are provided by price requesters and would consist of the list of launches, where each launch is described by

> - Launch id - Name of the SpaceX mission, that allows to uniquely identify particular launch.

Reviewer 4: ```suggestion
- Launch id - Name of the SpaceX mission, that allows voters to uniquely identify a particular launch.
```

- Weight (`Wi`) - Mathematical `weight` of the launch in the overall success rate calculation. Specified wight might be any positive number that is used to define the importance of this particular launch in the overall success rate respectively to other launches provided in ancillary data.

Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions

> - If launch did not happen until request date or the rocket was not successfully lifted off, then `Si=0`

Reviewer 4: ```suggestion
- If launch did not happen until request date or the rocket did not successfully lift off, then `Si=0`
```

- If the rocket successfully lifted off, but never landed, or the landing was not successful then `Si=0.5`
- If the rocket successfully lifted off, and landed, then `Si=1`

Liftoff is considered successful if the rocket reached expected altitude without any significant damage and performed the required part of the mission at this altitude.

Landing is considered successful if the rocket landed on expected surface without any significant damage until request date.

Success rate than can be calculated by
```
Success Rate =  SUM(Si * Wi) / SUM(Wi)
```

i.e. sum of all products of launch statuses and corresponding weights divided by the sum of all weights

This formula will return weighted average success rate on all specified launches in rage `[0, 1]`.

## Examples

### Example 1

- Request timestamp: 1614556800 (March 1st 2021 00:00 UTC)
- List of launches:
- - Launch 1: `{ "id": "Starlink-18", "w": 1 }`
- - Launch 2: `{ "id": "Starlink-19", "w": 1 }`

> Launch 1 happened until request date, successfully lifted off, and landed: `S1=1`

Reviewer 4: ```suggestion
Launch 1 happened before request date, successfully lifted off, and landed: `S1=1`
```

> Launch 2 happened until request date, successfully lifted off, but did not land: `S2=0.5`

Reviewer 4: ```suggestion
Launch 2 happened before request date, successfully lifted off, but did not land: `S2=0.5`
```

Calculating the success rate:
```
Success Rate = (1 * 1 + 1 * 0.5) / (1 + 1) = 0.75
```

### Example 2
- Request timestamp: 1612137600 (February 1st 2021 00:00 UTC)
- List of launches:
- -  Launch 1: `{ "id": "Transporter-1", "w": 1 }`

> Launch 1 happened until request date, successfully lifted off, and landed: `S1=1`

Reviewer 4: ```suggestion
Launch 1 happened before request date, successfully lifted off, and landed: `S1=1`
```

Calculating the success rate:
```
Success Rate = (1 * 1) / (1) = 1
```

## Format

> Ancillary data in the request will be passed as a JSON  in the following format

Reviewer 4: @Alirun can you also add some clarification here for a fallback in situations where ancillary data cannot be identified?

Something like:

```
If voters cannot correctly match a specific ID to an individual launch, `Si` should equal 0 for that launch.

If voters cannot correctly decode the ancillary data into the format defined below, voters should return 0.5.
```

Doesn't necessarily need to be 0.5, but we need some contingency instructions for situations where ancillary data is mis-parameterized.
